bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,206 @@
+                    expected_server_name: "localhost",
+                })
+                .unwrap(),
+        };
+
+        builder.build().unwrap()
+    }
+
+    /// Set up connections with config and custom IO
+    fn init_conn(&mut self, mode: Mode) -> Result<(), s2n_tls::error::Error> {
+        let client_to_server_ptr =
+            &mut self.client_to_server_buf as &mut VecDeque<u8> as *mut VecDeque<u8> as *mut c_void;
+        let server_to_client_ptr =
+            &mut self.server_to_client_buf as &mut VecDeque<u8> as *mut VecDeque<u8> as *mut c_void;
+        let (read_ptr, write_ptr, config, conn);
+
+        match mode {
+            Mode::Client => {
+                read_ptr = server_to_client_ptr;
+                write_ptr = client_to_server_ptr;
+                config = &self.client_config;
+                conn = &mut self.client_conn;
+            }
+            Mode::Server => {
+                read_ptr = client_to_server_ptr;
+                write_ptr = server_to_client_ptr;
+                config = &self.server_config;
+                conn = &mut self.server_conn;
+            }
+        }
+
+        conn.set_blinding(Blinding::SelfService)?
+            .set_config(config.clone())?
+            .set_send_callback(Some(Self::send_cb))?
+            .set_receive_callback(Some(Self::recv_cb))?;
+        unsafe {
+            conn.set_send_context(write_ptr)?
+                .set_receive_context(read_ptr)?;
+        }
+
+        Ok(())
+    }
+
+    /// Handshake step for one connection
+    fn handshake_conn(&mut self, mode: Mode) {
+        let (conn, handshaked);
+        match mode {
+            Mode::Client => {
+                debug!("Client: ");
+                conn = &mut self.client_conn;
+                handshaked = &mut self.client_handshaked;
+            }
+            Mode::Server => {
+                debug!("Server: ");
+                conn = &mut self.server_conn;
+                handshaked = &mut self.server_handshaked;
+            }
+        }
+        if let Ready(res) = conn.poll_negotiate() {
+            res.unwrap();
+            *handshaked = true;
+        } else {
+            *handshaked = false;
+        }
+    }
+}
+
+impl TlsBenchHarness for S2nTls {
+    fn new() -> Self {
+        debug!("----- s2n-tls -----");
+        let mut new_struct = S2nTls {
+            client_to_server_buf: Box::pin(VecDeque::new()),
+            server_to_client_buf: Box::pin(VecDeque::new()),
+            client_config: Self::create_config(Mode::Client),
+            server_config: Self::create_config(Mode::Server),
+            client_conn: Connection::new_client(),
+            server_conn: Connection::new_server(),
+            client_handshaked: false,
+            server_handshaked: false,
+        };
+        new_struct.init_conn(Mode::Client).unwrap();
+        new_struct.init_conn(Mode::Server).unwrap();
+        new_struct
+    }
+
+    fn handshake(&mut self) {
```

Comment:
thought: This function feels a little too "nice" to me. It checks if you've handshaked before, it uses a while loop to bail out early even if we were wrong on the number of loops, etc. I'd prefer for the code to handle that less gracefully, and force us to be more "correct" in our usage of this function.